locking/locks.go

```diff
@@ -218,100 +219,108 @@ func (c *Client) SearchLocks(filter map[string]string, limit int, localOnly bool
 			return []Lock{}, errors.New("can't search cached locks when filter or limit is set")
 		}
 
-		cacheFile, err := c.prepareCacheDirectory()
-		if err != nil {
-			return []Lock{}, err
-		}
-
-		_, err = os.Stat(cacheFile)
-		if err != nil {
-			if os.IsNotExist(err) {
-				return []Lock{}, errors.New("no cached locks present")
-			}
-
-			return []Lock{}, err
-		}
-
-		return c.readLocksFromCacheFile(cacheFile)
+		locks := []Lock{}
+		err := c.readLocksFromCacheFile("remote", func(decoder *json.Decoder) error {
+			return decoder.Decode(&locks)
+		})
+		return locks, err
 	} else {
 		locks, err := c.searchRemoteLocks(filter, limit)
 		if err != nil {
 			return locks, err
 		}
 
 		if len(filter) == 0 && limit == 0 {
-			cacheFile, err := c.prepareCacheDirectory()
-			if err != nil {
-				return locks, err
-			}
-
-			err = c.writeLocksToCacheFile(cacheFile, locks)
+			err = c.writeLocksToCacheFile("remote", func(writer io.Writer) error {
+				return c.EncodeLocks(locks, writer)
+			})
 		}
 
 		return locks, err
 	}
 }
 
-func (c *Client) VerifiableLocks(ref *git.Ref, limit int) (ourLocks, theirLocks []Lock, err error) {
-	if ref == nil {
-		ref = c.RemoteRef
-	}
-
+func (c *Client) SearchLocksVerifiable(limit int, cached bool) (ourLocks, theirLocks []Lock, err error) {
 	ourLocks = make([]Lock, 0, limit)
 	theirLocks = make([]Lock, 0, limit)
-	body := &lockVerifiableRequest{
-		Ref:   &lockRef{Name: ref.Refspec()},
-		Limit: limit,
-	}
 
-	c.cache.Clear()
+	if cached {
+		if limit != 0 {
+			return []Lock{}, []Lock{}, errors.New("can't search cached locks when limit is set")
```

Comment:
Is there a reason we use `[]Lock{}` instead of `nil`? I think it's customary to use `nil` as the zero value for slices when returning errors.

Reply:
It seems that the lock-related functions work differently here: they will always return results collected up to the point where the error has occurred. For instance, `VerifiableLocks` (before any of my changes from last year) return `ourLocks, theirLocks, err` with `ourLocks` and `theirLocks` being non-`nil` always:

https://github.com/git-lfs/git-lfs/blob/8b41ba5cb9bbff6564f2c75c8c5050cee17707ba/locking/locks.go#L238-L240

This somehow seems reasonable because this will give at least all results which could be collected until a (possible connection) error has occurred.

For the particular piece of code, we could change it to `nil, nil, errors.New(...)` but this requires to introduce some more checks in calling methods. Hence, I would prefer to leave as is.

Reply:
Okay, that seems fine.